.github/actions/preflight-k3d/preflight.sh

```
@@ -0,0 +1,152 @@
+# they are only free if the previous run's teardown actually ran, so check
+# explicitly and name the holder — quick-start/install.sh's own check reports
+# the port but not the process.
+# ---------------------------------------------------------------------------
+
+echo "::group::Required host ports"
+busy=""
+for p in ${REQUIRED_PORTS}; do
+  if holder="$(lsof -nP -iTCP:"${p}" -sTCP:LISTEN 2>/dev/null | tail -n +2)" && [ -n "${holder}" ]; then
+    busy="${busy} ${p}"
+    echo "  port ${p} IN USE:"
+    echo "${holder}" | sed 's/^/    /'
```

Shellcheck (0.11.0)

[style] See if you can use ${variable//search/replace} instead.

(SC2001)